integration/run_test.go

```diff
@@ -296,135 +287,77 @@ max_wait_msec = ` + strconv.FormatInt(tt.config.maxWaitMsec, 10) + `
 
-			m := rebootContainerd(t, sh, getContainerdConfigToml(t, false), getSnapshotterConfigToml(t, false, config))
+			rebootContainerd(t, sh, getContainerdConfigToml(t, false), getSnapshotterConfigToml(t, false, config))
 			// Mirror image
 			copyImage(sh, dockerhub(containerImage), regConfig.mirror(containerImage))
-			// Pull image, create SOCI index with all layers and small (100kiB) spans
-			indexDigest := buildIndex(sh, regConfig.mirror(containerImage), withMinLayerSize(0), withSpanSize(100*1024))
-			sh.X(append(imagePullCmd, "--soci-index-digest", indexDigest, regConfig.mirror(containerImage).ref)...)
-
-			// Run the container
 			image := regConfig.mirror(containerImage).ref
-			sh.X(append(runSociCmd, "--name", "test-container", "-d", image, "sleep", "infinity")...)
-
-			sh.X("apt-get", "--no-install-recommends", "install", "-qy", "iptables")
-
-			// TODO: Wait for the container to be up and running
+			indexDigest := buildIndex(sh, regConfig.mirror(containerImage), withMinLayerSize(0), withSpanSize(1<<20))
+			sh.X("soci", "push", "--user", regConfig.creds(), regConfig.mirror(containerImage).ref)
 
+			rebootContainerd(t, sh, getContainerdConfigToml(t, false), getSnapshotterConfigToml(t, false, config))
+			// Re-pull image from our local registry mirror
+			sh.X(append(imagePullCmd, "--soci-index-digest", indexDigest, regConfig.mirror(containerImage).ref)...)
+			sh.X("apt-get", "-qq", "--no-install-recommends", "install", "-y", "iptables")
 			// Block network access to the registry
 			if tt.config.networkDisableMsec > 0 {
 				sh.X("iptables", "-A", "OUTPUT", "-d", registryHostIP, "-j", "DROP")
 			}
 
-			// Do something in the container that should work without network access
-			commandSucceedStdout, _, err := sh.R("nerdctl", "exec", "test-container", "sh", "-c", "times")
+			// Read a file to trigger a synchronous network request.
+			cmdStdout, cmdStderr, err := sh.R(append(runSociCmd, image, "cat", "/etc/hosts")...)
 			if err != nil {
-				t.Fatalf("attempt to run task without network access failed: %s", err)
-			}
-
-			type ErrorLogLine struct {
-				Error string `json:"error"`
-				Msg   string `json:"msg"`
+				t.Fatalf("attempt to run task requiring network access failed: %s", err)
 			}
 
-			gaveUpChannel := make(chan bool, 1)
-			defer close(gaveUpChannel)
+			successChannel := make(chan string, 1)
+			failureChannel := make(chan string, 1)
 
-			monitorGaveUp := func(rawL string) {
-				if i := strings.Index(rawL, "{"); i > 0 {
-					rawL = rawL[i:] // trim garbage chars; expects "{...}"-styled JSON log
-				}
-				var logLine ErrorLogLine
-				if err := json.Unmarshal([]byte(rawL), &logLine); err == nil {
-					if logLine.Msg == "statFile error" && strings.Contains(logLine.Error, "giving up after") {
-						gaveUpChannel <- true
-						return
-					}
+			listener := func(c chan string) func(string) {
+				return func(rawLog string) {
+					var once sync.Once
+					once.Do(func() { c <- rawLog })
 				}
 			}
```

Comment:
You would need to move this higher in scope else a new variable will be initialized when the listener return function is called. i.e.
```suggestion
            listener := func(c chan string) func(string) {
                    var once sync.Once
                    return func(rawLog string) {
                        once.Do(func() { c <- rawLog })
                    }
            }
```

Reply:
The intention here is to allow writing to both stdout and stderr in a single test case. Then, in the [select block](https://github.com/awslabs/soci-snapshotter/pull/1175/files#diff-8a6db7c58ab40add58f9046cee9ffab883d23e66c3dbc4a315dba43314c06d0bR350) further down, if one is written to unintentionally, it would get caught here and return an error.

Reply:
Just pushed a change to make the latter checks more rigorous. Hopefully it clarifies the intention better.

Reply:
If I understand this correctly, Austin's concern is that the `sync.Once` is a new instance for every log line instead of being a single instance across all log lines.

Reply:
In that case, would the solution still work? Under the assumption this is pass-by-value, then it definitely should, but I'm not sure of a way to confirm this.

Reply:
idk why I said that, I just added a few print statements to check.

Anything before the return statement always runs. Anything within it is run every time we get a new message. So Austin was correct. Made the changes Austin recommended.